library/Haste/Dca/AjaxOperations.php

```diff
@@ -0,0 +1,303 @@
+                $id
+            ),
+            __METHOD__,
+            TL_GENERAL
+        );
+        
+        $response = array(
+            'nextValue' => $options[$nextIndex]['value'],
+            'nextIcon'  => $options[$nextIndex]['icon']
+        );
+
+        $response = new JsonResponse($response);
+        $response->send();
+    }
+
+    /**
+     * Modifies the DCA.
+     *
+     * @param string $table
+     */
+    public function modifyDca($table)
+    {
+        if (TL_MODE !== 'BE'
+            || !isset($GLOBALS['TL_DCA'][$table]['list']['operations'])
+            || !is_array($GLOBALS['TL_DCA'][$table]['list']['operations'])
+        ) {
+            return;
+        }
+
+        foreach ($GLOBALS['TL_DCA'][$table]['list']['operations'] as $name => $settings) {
+            if (!isset($settings['haste_ajax_operation'])) {
+                continue;
+            }
+
+            $operation = &$GLOBALS['TL_DCA'][$table]['list']['operations'][$name];
+
+            // Add the JavaScript
+            $GLOBALS['TL_JAVASCRIPT'][] = Debug::uncompressedFile('system/modules/haste/assets/haste.min.js') . '|static';
+
+            // Add default button callback to display the correct initial state
+            // but only add it if not already present
+            if (!isset($settings['button_callback'])) {
+                $operation['button_callback'] = $this->getDefaultButtonCallback($name, $table, $settings['haste_ajax_operation']);
+
+                // Make sure an icon is set to prevent DC_Table errors
+                // (set to '' as the button_callback will return the correct icon)
+                $operation['icon'] = '';
+
+                // Add the onclick attribute
+                $this->addOnClickAttribute($operation);
+            }
+        }
+    }
+
+    /**
+     * Adds the "onclick" attribute to the operation DCA.
+     *
+     * @param $operation
+     */
+    private function addOnClickAttribute(&$operation)
+    {
+        $clickEventString = 'return Haste.toggleAjaxOperation(this, %s);';
+
+        if (!isset($operation['attributes'])) {
+            $operation['attributes'] = sprintf('onclick="%s"', $clickEventString);
+        } else {
+            // onclick attribute already present
+            if (strpos($operation['attributes'], 'onclick="') !== false) {
+                $operation['attributes'] = str_replace(
+                    'onclick="',
+                    'onclick="' . $clickEventString,
+                    $operation['attributes']
+                );
+            } else {
+                $operation['attributes'] = $clickEventString . $operation['attributes'];
+            }
+        }
+    }
+
+    /**
+     * Checks if user has the permissions for the field.
+     *
+     * @param       $table
+     * @param array $hasteAjaxOperationSettings
+     *
+     * @return bool
+     */
+    private function checkPermission($table, array $hasteAjaxOperationSettings)
```

Comment:
Generally I think it would be better not to pass the config array but to pass the necessary values directly. This way the methods don't need to know about the config array setup. Applies to other methods, too